### DISCUSSION TITLE
MQTT on QUIC

### DISCUSSION THREAD

**Discussion by zmstone** [emqx/emqx#4379] [Show and tell]
# NIF PoC [status: completed]

Exploring:

OS: Linux

-   NIF compilation, debug/non-debug build

-   Resource planning
    OS Thread modeling
    
    CPU / Memory usage.
    
    Shared Data (Erlang VM and MsQuic worker threads)
    
    OS resources: Sockets, buffers

-   simple client/server works via the loopback device

-   find out feature limitations.

-   hardcode configurations.

# Dev pipelines & Useability 

-   Provides inet and gen<sub>tcp</sub> like APIs.
    
    Happy path, very few error handlings.
    
    with basic options but skip fine-tuning options:
    
    basic options:
      remote IP, remote Port, Path to SSL cert file&#x2026;
    
    fine-tuning options:
      buffer size, queue size, timeouts &#x2026;.
    
    None-protocol configurations are skipped:
      MSQuick stack specific configuration.

-   Provide basic debug / tracing methods.

-   Dev CI pipeline
    basic tests for API above.
    Atomic result: Pass/Fail.
    Gatekeeper for PRs on the Github.

-   Build on macOS

# MQTT product integration [inprogress]

-   Integration with the client: emqtt

-   Integration with the server: emqx

-   Demo: MQTT client/server over emqtt

-   Testcases are provided in emqtt and emqx.

# More Error handling and Strengthen the test

-   Bring in property-based test

-   Bring in trace-based test

-   Bring in googletest for Nif code

-   Bring in code static analysis tools

# Able to run Load test and robustness test

-   Perf. Monitoring of cluster.

-   Load test env provision.

-   Generate load test reports.

# Load /robustness improvements

# MQTT on QUIC production ready (Unix/Linux)

**Comment by zmstone** [comment]
Project lead: @qzhuyan

**Comment by Gypsying** [comment]
Can emqx-5.0.x support `QUIC IETF: draft-29 (0xff00001d)` ?
> What [document](https://www.emqx.io/docs/en/v5.0/advanced/quic.html) described is [RFC9000](https://datatracker.ietf.org/doc/html/rfc9000).

I used golang program(base draft-29 protocol) to connet emqx-5.0.4, 
the connection is created but client side always said :"`[Expert Info (Note/Protocol): Unknown QUIC connection. Missing Initial Packet or migrated connection?]`".
![image](https://user-images.githubusercontent.com/17995262/185024630-c1f52c58-23b6-4a6a-b0bf-ba0731fbda8f.png)


> There is no standard go sdk to connect mqtt over quic, i just use the [hack version](https://pdfs.semanticscholar.org/9c86/6f66ea40cce43518bbf81ed670121c671726.pdf) to tested.

I'm confused, any idea about this? @qzhuyan

**Reply by qzhuyan** [reply]
hi,
It should work with draft-29.
I need to see the entire quic trace, in order to tell.

Do you put `mqtt` in alpn list while doing the handshake? 

ref: https://www.emqx.io/docs/en/v5/advanced/quic.html#implementation

**Reply by Gypsying** [reply]
> hi, It should work with draft-29. I need to see the entire quic trace, in order to tell.
> 
> Do you put `mqtt` in alpn list while doing the handshake?
> 
> ref: https://www.emqx.io/docs/en/v5/advanced/quic.html#implementation

Yes, i put the mqtt to alpn list, such as:
![image](https://user-images.githubusercontent.com/17995262/190544118-86cb096c-ee97-470f-83aa-721c4527acde.png)

**Comment by Gypsying** [comment]
I reproduce the [error](https://github.com/emqx/emqx/discussions/4379#discussioncomment-3411111) with RFC9000.
So this is not caused by quic version. @qzhuyan 

I will look for the real reason...

**Reply by qzhuyan** [reply]
@Gypsying It will be helpful for us to troubleshoot if you could send us a tcpdump/wireshake dump.
